mqtt_io/server.py

```diff
@@ -1191,16 +1169,48 @@ async def run() -> None:
                 try:
                     await asyncio.gather(*self.critical_tasks)
                 except asyncio.CancelledError:
-                    pass
+                    break
                 except Exception:  # pylint: disable=broad-except
                     _LOG.exception("Exception in critical task:")
             except asyncio.CancelledError:
-                pass
+                break
+            except MQTTException:
+                if counter > 0:
+                    counter -= 1
+                _LOG.exception('Connection to MQTT-Broker failed')
+                if counter == 0:
+                    break
```

Comment:
Since we've already decremented `counter`, won't this mean that when `reconnect_count` is set to 0 or 1, we won't try reconnecting at all? If it's set to 2, we'll only reconnect once and so on?

Reply:
Yes, you were right. I've fixed that now.